presto-main/src/main/java/com/facebook/presto/sql/analyzer/GroupingOperationAnalyzer.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.DefaultExpressionTraversalVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+public class GroupingOperationAnalyzer
```

Comment:
So I implemented something but then realized that because of the Visitor pattern I don't think what you guys envision is going to be possible. Here's the class I wrote:
```
public class ExpressionExtractor
        extends DefaultExpressionTraversalVisitor<Void, Void>
{
    private final ImmutableList.Builder<Expression> expressionBuilder = ImmutableList.builder();
    private final Function<Expression, Boolean> predicate;

    public ExpressionExtractor(Function<Expression, Boolean> predicate)
    {
        this.predicate = requireNonNull(predicate, "predicate is null");
    }

    @Override
    protected Void visitExpression(Expression node, Void context)
    {
        if (predicate.apply(node)) {
            expressionBuilder.add(node);
        }

        return null;
    }

    public List<Expression> getExpressions()
    {
        return expressionBuilder.build();
    }
}
```
With the following invocation for extracting aggregates:

```
        ExpressionExtractor aggregateExtractor = new ExpressionExtractor(
                (expression) -> {
                    if (expression instanceof FunctionCall) {
                        FunctionCall functionCall = (FunctionCall) expression;
                        if ((functionRegistry.isAggregationFunction(functionCall.getName()) || functionCall.getFilter().isPresent()) && !functionCall.getWindow().isPresent()) {
                            return true;
                        }
                    }
                    return false;
                });
        aggregateExtractor.process(predicate, null);
```

The reason why the above doesn't work is because if the input expression is a `FunctionCall` then the `ExpressionExtractor#visitExpression()` is never going to be called and instead `DefaultExpressionTraversalVisitor#visitFunctionCall()` is going to be called and by the time we get to `visitExpression()` the original/parent expression is lost and we're left with looking at the arguments of the `FunctionCall`.

One possibility is to basically consolidate all three extractors into one class but in that class override `visitFunctionCall()` and `visitGroupingOperation()`. However, every time someone wants to be able to extract a new type of expression they'll have to override the corresponding method.

What do you think @kokosing and @sopel39?